design/open/delta-diff.md

```diff
@@ -0,0 +1,186 @@
+as long as it's self-contained (theoretically it can also be not system-runtime-dependent but then the cost will be an added requirement for running lakeFS- runtime support).
+
+#### Plugin system consideration
+
+* The plugin system should be flexible enough such that it won't impose a language restriction, so that in the future we could have lakeFS users take part in plugins creation (not restricted to diffs).
+* The plugin system should support multiple OSs.
+* The plugin system should be easy enough to write plugins.
+
+#### Hashicorp Plugin system
+
+Hashicorp's battle-tested [`go-plugin`](https://github.com/hashicorp/go-plugin) system is a plugin system over RPC (used in Terraform, Vault, Consul, and Packer).
+Currently, it's only designed to work over a local network.  
+Plugins can be written and consumed in almost every major language. This is achieved by supporting [gRPC](http://www.grpc.io/) as the communication protocol.
+* The plugin system works by launching subprocesses and communicating over RPC (both `net/rpc` and `gRPC` are supported).
+  A single connection is made between any plugin and the core process. For gRPC-based plugins, the HTTP2 protocol handles [connection multiplexing](https://freecontent.manning.com/animation-http-1-1-vs-http-2-vs-http-2-with-push/).
+* The plugin and core system are separate processes, which means that a crash of a plugin won't cause the core system to crash (lakeFS in our case).
+
+![Microkernel architecture overview](diagrams/microkernel-overview.png)
+[(excalidraw file)](diagrams/microkernel-overview.excalidraw)
+
+---
+
+## Implementation
+
+![Delta Diff flow](diagrams/delta-diff-flow.png)
+[(excalidraw file)](diagrams/delta-diff-flow.excalidraw)
+
+#### DiffService
+
+The DiffService will be an internal component in lakeFS which will serve as the Core system.  
+In order to realize which diff plugins are available, we shall use new configuration values:
+1. `LAKEFS_PLUGINS_LOCATION`
+   The DiffService will use this location to search for the required plugin. Defaults to lakeFS's configuration file directory.
+2. `LAKEFS_PLUGIN_DIFF_{TYPE OF DIFF}`
+   The DiffService will use the name of binary provided by this env var to perform the diff. For instance, `LAKEFS_PLUGIN_DIFF_DELTA=deltaDiffBinary`
+   The DiffService will use the [plugins' location](https://github.com/hashicorp/terraform/blob/main/plugins.go) and the diff binary to load the plugin and request a diff.
+   The **type** of diff will be sent as part of the request to lakeFS as specified [here](#API).
+   The communication between the DiffService and the plugins, as explained above, will be through the `go-plugin` package (`gRPC`).
+
+#### Delta Diff Plugin
+
+Implemented using [delta-rs](https://github.com/delta-io/delta-rs) (Rust), this plugin will perform the diff operation using table paths provided by the DiffService through a `gRPC` call.  
```

Comment:
- We should specify the constraint of fetching large objects from delta-rs. Configurable (optional) location to store the temporary data downloaded by the plugin - will probably be required in case we don't want to use the default temporary folder or control/strict the size used by the plugin.
- Cleanup or this storage if it is not managed by the OS temp dir/files API.
- Calling the plugin with 'deadline' (timeout for the grpc call) just to prevent a lock during the call to the plugin.